### PR TITLE
Add the bound check for flatten with out_dim

### LINKED ISSUE #120762: Segmentation Fault due to Invalid out_dim Type in `torch.tensor.flatten`
### 🐛 Describe the bug

When attempting to flatten a tensor using PyTorch's `torch.tensor.flatten` method with the `out_dim` parameter set to a string value, it triggers a segmentation fault resulting in a core dump. 

## Error log
```bash
munmap_chunk(): invalid pointer

or

malloc(): unaligned tcache chunk detected
Aborted (core dumped)
```

## Minified Repro
```python
import torch

a = torch.tensor([1, 2, 3])
a.flatten(start_dim=0, end_dim=1, out_dim='a')

```

### Versions

PyTorch version: 2.3.0.dev20240220+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.8.17 (default, Jul  5 2023, 21:04:15)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-15-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      48 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             256
On-line CPU(s) list:                0-255
Vendor ID:                          AuthenticAMD
Model name:                         AMD EPYC 7763 64-Core Processor
CPU family:                         25
Model:                              1
Thread(s) per core:                 2
Core(s) per socket:                 64
Socket(s):                          2
Stepping:                           1
Frequency boost:                    enabled
CPU max MHz:                        3529.0520
CPU min MHz:                        1500.0000
BogoMIPS:                           4890.76
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                     AMD-V
L1d cache:                          4 MiB (128 instances)
L1i cache:                          4 MiB (128 instances)
L2 cache:                           64 MiB (128 instances)
L3 cache:                           512 MiB (16 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-63,128-191
NUMA node1 CPU(s):                  64-127,192-255
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240220+cu118
[pip3] torchaudio==2.2.0.dev20240220+cu118
[pip3] torchvision==0.18.0.dev20240220+cu118
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-triton            3.0.0+901819d2b6          pypi_0    pypi
[conda] torch                     2.3.0.dev20240220+cu118          pypi_0    pypi
[conda] torchaudio                2.2.0.dev20240220+cu118          pypi_0    pypi
[conda] torchvision               0.18.0.dev20240220+cu118          pypi_0    pypi

cc @malfet

### PR DESCRIPTION
**Pull request by lancerts**
Fixes #120762

The bound is not valid in the example but unchecked.
```
a = torch.tensor([1, 2, 3])
a.flatten(start_dim=0, end_dim=1, out_dim='a')
```

The same is checked for the case 

```
a = torch.tensor([1, 2, 3])
a.flatten(start_dim=0, end_dim=1)
```

- Therefore, just apply the same check.

@malfet @janeyx99

### CODE REVIEW

**Review by spzala — COMMENTED** [review]
Thanks @lancerts !! I think we should cover this check in the existing flatten test cases as well.

**Comment by lancerts** [comment]
> Thanks @lancerts !! I think we should cover this check in the existing flatten test cases as well.

@spzala added the test cases. ty

File: test/test_namedtensor.py [hunk 1]
```diff
@@ -1077,6 +1077,21 @@ def test_flatten_nodims(self):
         with self.assertRaisesRegex(RuntimeError, "cannot be empty"):
             tensor.flatten((), 'abcd')
 
+    def test_flatten_index_error(self):
+        tensor = torch.randn(1, 2)
+        with self.assertRaisesRegex(IndexError,
+                                    "IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)"):
```
Comment (malfet) [inline test/test_namedtensor.py:1083]:
```suggestion
                                    "Dimension out of range (expected to be in range of [-2, 1], but got 2)"):
```

File: test/test_namedtensor.py [hunk 2]
```diff
@@ -1077,6 +1077,21 @@ def test_flatten_nodims(self):
...
+            tensor.flatten(0, 2)
+        with self.assertRaisesRegex(IndexError,
+                                    "IndexError: Dimension out of range (expected to be in range of [-2, 1], but got 2)"):
+            tensor.flatten(0, 2, 'N')
+        with self.assertRaisesRegex(RuntimeError,
+                                    "RuntimeError: flatten() has invalid args: start_dim cannot come after end_dim"):
```
Comment (malfet) [inline test/test_namedtensor.py:1089]:
```suggestion
                                    "flatten() has invalid args: start_dim cannot come after end_dim"):
```

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -3419,6 +3419,10 @@ Tensor flatten(const Tensor& self, int64_t start_dim, int64_t end_dim) {
 }
 
 Tensor flatten(const Tensor& self, int64_t start_dim, int64_t end_dim, Dimname out_dim) {
+  start_dim = maybe_wrap_dim(start_dim, self.dim());
+  end_dim = maybe_wrap_dim(end_dim, self.dim());
+  TORCH_CHECK(start_dim <= end_dim, "flatten() has invalid args: start_dim cannot come after end_dim");
```
Comment (malfet) [inline aten/src/ATen/native/TensorShape.cpp:3424]: Hmm, shouldn't it be a ValueError or and IndexError? Also, please mention start and end dim in the error message

**Review by malfet — APPROVED** [review]
LGTM

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (lancerts) [reply aten/src/ATen/native/TensorShape.cpp:3424]: Thanks for the suggestion. I think it should be an IndexError or ValueError indeed.

I took the check from the basecase (without the out_dim arg)

https://github.com/pytorch/pytorch/blob/33da8d5c1233e0cedeb90a97ffcd360102f667b7/aten/src/ATen/native/TensorShape.cpp#L3394


I can change it to ValueError with start/end dim in the error message but for consistency, need to also change the one from the basecase.

@malfet Should i go ahead for the change at both places or leave things as is? Thanks

Comment (lancerts) [reply aten/src/ATen/native/TensorShape.cpp:3424]: same TORCH_CHECK changes required here if go ahead

https://github.com/pytorch/pytorch/blob/33da8d5c1233e0cedeb90a97ffcd360102f667b7/aten/src/ATen/native/metal/ops/MetalReshape.mm#L77

Comment (malfet) [reply aten/src/ATen/native/TensorShape.cpp:3424]: Hmm, let's leave it until followups then, because changing those might be considered BC breaking

Comment (lancerts) [reply aten/src/ATen/native/TensorShape.cpp:3424]: Sure, can you re-check if everything looks good (commited the suggested change)? Thanks @malfet

Comment (spzala) [reply aten/src/ATen/native/TensorShape.cpp:3424]: Probably a good idea to open an issue for the TORCH_CHECK level changes for tracking it separately? Thanks!

**Review by spzala — APPROVED** [review]
LGTM. Thanks @lancerts !!

**Comment by spzala** [comment]
> @pytorchbot merge

@lancerts please wait for @malfet approval before attempting to merge. Thanks!

**Review by malfet — APPROVED** [review]

**Comment by lancerts** [comment]
> ## Merge failed
> **Reason**: 3 mandatory check(s) failed. The first few are:
> 
> * [BC Lint](https://github.com/pytorch/pytorch/actions/runs/8104738453)
> * [Lint](https://github.com/pytorch/pytorch/actions/runs/8104738520)
> * [pull](https://github.com/pytorch/pytorch/actions/runs/8104743701)
> 
> Dig deeper by [viewing the failures on hud](https://hud.pytorch.org/pytorch/pytorch/commit/d01f60e119b0b903e9f780fa421232956efef23c)
> 
> Details for Dev Infra team

@malfet sorry for the ping, do we need additional approval?

**Comment by lancerts** [comment]
@malfet fixed the lint error and proper escape for the error message in the test case, thanks

**Pull request closed without merging**